Blog Post on Flink 1.10 Python UDF support V2 (#316)
apache/flink-web: _posts/2020-02-23-pyflink-udf-support-flink.md

---
layout: post
title: "PyFlink: Introducing Python Support for UDFs in Flink's Table API"
date: 2020-02-23T12:00:00.000Z
authors:
- Jincheng:
  name: "Jincheng Sun"
  twitter: "sunjincheng121"
- markos:
  name: "Markos Sfikas"
  twitter: "MarkSfik"
excerpt: Flink 1.10 extends its support for Python by adding Python UDFs in PyFlink. This post explains how UDFs work in PyFlink and gives some practical examples of how to use UDFs in PyFlink. 
---

Flink 1.9 introduced the Python Table API, allowing developers and data engineers to write Python Table API jobs for Table transformations and analysis, such as Python ETL or aggregate jobs. However, Python users faced some limitations when it came to support for Python UDFs in Flink 1.9, preventing them from extending the system’s built-in functionality. 

In Flink 1.10, the community further extended the support for Python by adding Python UDFs in PyFlink. Additionally, both the Python UDF environment and dependency management are now supported, allowing users to import third-party libraries in the UDFs, leveraging Python's rich set of third-party libraries. 

# Python Support for UDFs in Flink 1.10

> Before diving into how you can define and use Python UDFs, we explain the motivation and background behind how UDFs work in PyFlink and provide some additional context about the implementation of our approach. Below we will give a brief introduction on the PyFlink architecture from job submission to executing the Python UDF.

Reviewer: ```suggestion
Before diving into how you can define and use Python UDFs, we explain the motivation and background behind how UDFs work in PyFlink and provide some additional context about the implementation of our approach. Below we give a brief introduction on the PyFlink architecture from job submission, all the way to executing the Python UDF.
```

> The architecture mainly includes two parts, local and cluster, as shown in architecture picture below. The local phase is the compilation of the job, and the cluster is the execution of the job.

Reviewer: ```suggestion
The PyFlink architecture mainly includes two parts — local and cluster — as shown in the architecture visual below. The local phase is the compilation of the job, and the cluster is the execution of the job.
```

<center>
<img src="{{ site.baseurl }}/img/blog/2020-02-23-pyflink-udfs/pyflink-udf-architecture.png" width="600px" alt="PyFlink UDF Architecture"/>
</center>
<br>

> For the local part, the Python API is a mapping of the Java API, that is, each time Python executes a method in the figure above, it will synchronously call the method corresponding to Java through Py4J, and finally generate a Java JobGraph, and then submit it to the cluster.

Reviewer: ```suggestion
For the local part, the Python API is a mapping of the Java API, that is, each time Python executes a method in the figure above, it will synchronously call the method corresponding to Java through Py4J, and finally generate a Java JobGraph, before submiting it to the cluster.
```

> For the cluster part, just like ordinary Java jobs, the JobMaster schedules tasks to TaskMangers. The tasks that include Python UDF in a TaskManager involve the execution of Java and Python operators. In the Python UDF operator, various gRPC services are used to provide different communications between the Java VM and the Python VM, such as DataService for data transmissions, StateService for state requirements, and Logging and Metrics Services. These services are built on Beam's Fn API. Furthermore, it should be noted that although only Process mode is supported for Python workers now, Docker mode and External service mode are also considered to be supported later.

Reviewer: ```suggestion
For the cluster part, just like ordinary Java jobs, the JobMaster schedules tasks to TaskManagers. The tasks that include Python UDF in a TaskManager involve the execution of Java and Python operators. In the Python UDF operator, various gRPC services are used to provide different communications between the Java VM and the Python VM, such as DataService for data transmissions, StateService for state requirements, and Logging and Metrics Services. These services are built on Beam's Fn API. While currently only Process mode is supported for Python workers, support for Docker mode and External service mode is also considered for future Flink releases.
```

Suppose you want to use the `mpmath` to perform the sum of the example above. The Python UDF may look like:

```python
@udf(input_types=[DataTypes.BIGINT(), DataTypes.BIGINT()], result_type=DataTypes.BIGINT())
def add(i, j):
    from mpmath import fadd # add third-party dependency
    return int(fadd(1, 2))
```

To make it available on the worker node that does not contain the dependency, you can specify the dependencies with the following API:

```python
# echo mpmath==1.1.0 > requirements.txt
# pip download -d cached_dir -r requirements.txt --no-binary :all:
t_env.set_python_requirements("/path/of/requirements.txt", "/path/of/cached_dir")
```

A `requirements.txt` file that defines the third-party dependencies is used. If the dependencies cannot be accessed in the cluster, then you can specify a directory containing the installation packages of these dependencies by using the parameter "`requirements_cached_dir`", as illustrated in the example above. The dependencies will be uploaded to the cluster and installed offline. 

# Conclusion & Upcoming work

> In this blog post, we introduced the architecture of Python UDFs in PyFlink and provided some examples on how to define, register and invoke UDFs. Flink 1.10 brings Python support in the framework to new levels, allowing Python users to write even more magic with their preferred language. The community is actively working towards continuously improving the functionality and performance of PyFlink. Future work in upcoming releases will introduce support for Pandas UDFs in scalar and aggregate functions, add support to use Python UDFs through the SQL client to further expand the usage scope of Python UDFs, add support for Python ML Pipeline API and finally work towards even more performance improvements. More details can be found in the roadmap picture below. 

Reviewer: ```suggestion
In this blog post, we introduced the architecture of Python UDFs in PyFlink and provided some examples on how to define, register and invoke UDFs. Flink 1.10 brings Python support in the framework to new levels, allowing Python users to write even more magic with their preferred language. The community is actively working towards continuously improving the functionality and performance of PyFlink. Future work in upcoming releases will introduce support for Pandas UDFs in scalar and aggregate functions, add support to use Python UDFs through the SQL client to further expand the usage scope of Python UDFs, provide support for a Python ML Pipeline API and finally work towards even more performance improvements. The picture below provides more details on the roadmap for succeeding releases. 
```
